src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseUtilService.java

```diff
@@ -0,0 +1,557 @@
+        assertThat(tests).as("test case is initialized").hasSize(3);
+    }
+
+    public void addBuildPlanAndSecretToProgrammingExercise(ProgrammingExercise programmingExercise, String buildPlan) {
+        buildPlanRepository.setBuildPlanForExercise(buildPlan, programmingExercise);
+        programmingExercise.generateAndSetBuildPlanAccessSecret();
+        programmingExercise = programmingExerciseRepository.save(programmingExercise);
+
+        var buildPlanOptional = buildPlanRepository.findByProgrammingExercises_IdWithProgrammingExercises(programmingExercise.getId());
+        assertThat(buildPlanOptional).isPresent();
+        assertThat(buildPlanOptional.get().getBuildPlan()).as("build plan is set").isNotNull();
+        assertThat(programmingExercise.getBuildPlanAccessSecret()).as("build plan access secret is set").isNotNull();
+    }
+
+    public AuxiliaryRepository addAuxiliaryRepositoryToExercise(ProgrammingExercise programmingExercise) {
+        AuxiliaryRepository repository = new AuxiliaryRepository();
+        repository.setName("auxrepo");
+        repository.setDescription("Description");
+        repository.setCheckoutDirectory("assignment/src");
+        repository = auxiliaryRepositoryRepository.save(repository);
+        programmingExercise.setAuxiliaryRepositories(List.of(repository));
+        repository.setExercise(programmingExercise);
+        programmingExerciseRepository.save(programmingExercise);
+        return repository;
+    }
+
+    public void addSubmissionPolicyToExercise(SubmissionPolicy policy, ProgrammingExercise programmingExercise) {
+        policy = submissionPolicyRepository.save(policy);
+        programmingExercise.setSubmissionPolicy(policy);
+        programmingExerciseRepository.save(programmingExercise);
+    }
+
+    public ProgrammingSubmission addProgrammingSubmission(ProgrammingExercise exercise, ProgrammingSubmission submission, String login) {
+        StudentParticipation participation = participationUtilService.addStudentParticipationForProgrammingExercise(exercise, login);
+        submission.setParticipation(participation);
+        submission = programmingSubmissionRepo.save(submission);
+        return submission;
+    }
+
+    /**
+     * Add a submission with a result to the given programming exercise. The submission will be assigned to the corresponding participation of the given login (if exists or
+     * create a new participation).
+     * The method will make sure that all necessary entities are connected.
+     *
+     * @param exercise   for which to create the submission/participation/result combination.
+     * @param submission to use for adding to the exercise/participation/result.
+     * @param login      of the user to identify the corresponding student participation.
+     */
+    public void addProgrammingSubmissionWithResult(ProgrammingExercise exercise, ProgrammingSubmission submission, String login) {
```

Comment:
I am not sure if this method should be here (ProgrammingExerciseUtilService) or in the ParticipationUtilService. There are some similar methods in the ParticipationUtilService

Reply:
Discussed:
Methods related to submissions and specific Exercises go into the corresponding ExerciseUtilServices.
Submissions with no specific Exercise belong into ParticipationUtilService